sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,426 @@
+        gst::Element::link_many(&[&q, &conv, &sink]).unwrap();
+    }
+    let qpad = q.get_static_pad("sink").unwrap();
+    let ret = pad.link(&qpad);
+    assert_eq!(ret, gst::PadLinkReturn::Ok);
+}
+
+fn on_incoming_decodebin_stream(
+    values: &[glib::Value],
+    pipe: &gst::Element,
+) -> Option<glib::Value> {
+    let pad = values[1].get::<gst::Pad>().expect("Invalid argument");
+    if !pad.has_current_caps() {
+        println!("Pad {:?} has no caps, can't do anything, ignoring", pad);
+    }
+
+    let caps = pad.get_current_caps().unwrap();
+    let name = caps.get_structure(0).unwrap().get_name();
+    if name.starts_with("video") {
+        handle_media_stream(&pad, &pipe, "videoconvert", "autovideosink");
+    } else if name.starts_with("audio") {
+        handle_media_stream(&pad, &pipe, "audioconvert", "autoaudiosink");
+    } else {
+        println!("Unknown pad {:?}, ignoring", pad);
+    }
+    None
+}
+
+fn on_incoming_stream(values: &[glib::Value], pipe: &gst::Element) -> Option<glib::Value> {
+    let webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let decodebin = gst::ElementFactory::make("decodebin", None).unwrap();
+    let pipe_clone = pipe.clone();
+    decodebin
+        .connect("pad-added", false, move |values| {
+            on_incoming_decodebin_stream(values, &pipe_clone)
+        })
+        .unwrap();
+    pipe.clone()
+        .dynamic_cast::<gst::Bin>()
+        .unwrap()
+        .add(&decodebin)
+        .unwrap();
+    decodebin.sync_state_with_parent().unwrap();
+    webrtc.link(&decodebin).unwrap();
+    None
+}
+
+fn send_ice_candidate_message(
+    app_state: &Arc<AtomicUsize>,
+    values: &[glib::Value],
+    out: &ws::Sender,
+) -> Option<glib::Value> {
+    if app_state.load(Ordering::Relaxed) < (AppState::PeerCallNegotiating as usize) {
+        panic!("Can't send ICE, not in call");
+    }
+
+    let _webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let mlineindex = values[1].get::<u32>().expect("Invalid argument");
+    let candidate = values[2].get::<String>().expect("Invalid argument");
+    let message = json!({
+          "ice": {
+            "candidate": candidate,
+            "sdpMLineIndex": mlineindex,
+          }
+        });
+    out.send(message.to_string()).unwrap();
+    None
+}
+
+fn start_pipeline(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> gst::Element {
+    let pipe = gst::parse_launch(
+        "webrtcbin name=sendrecv
+        stun-server=stun://stun.l.google.com:19302
+        videotestsrc pattern=ball ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay !
+        queue !
+        application/x-rtp,media=video,encoding-name=VP8,payload=96 ! sendrecv.
+        audiotestsrc wave=red-noise ! audioconvert ! audioresample ! queue ! opusenc ! rtpopuspay !
+        queue !
+        application/x-rtp,media=audio,encoding-name=OPUS,payload=97 ! sendrecv.
+        ",
+    ).unwrap();
```

Comment:
Explicitly creating the elements would seem nicer than string programming, but I guess it's fine :)